### PR TITLE
Feature/permutation feature importance

### PR DESCRIPTION
**Pull request by miguelmartin75**
Implementation of blackbox feature permutation importance algorithm: https://christophm.github.io/interpretable-ml-book/feature-importance.html

Pseudo-code of the above, for a single batch of data:

```
perm_feature_importance(batch):
     feature_importance = {}
     initial_err = f(batch)
     for feature in features:
           batch_copy = permute_feature_across(batch, feature)
           err = f(batch_copy)
           diff = initial_err - err
           feature_importance[feature] = diff

     return feature_importance
           
permute_feature_across(batch, feature):
      batch_copy = batch.clone()
      batch_copy[:, feature] = batch[random_perm(n), feature] 
      return batch_copy
```

The basic idea is to permute each feature in a batch and see how different the output (or error/loss) is from the output/error/loss of the original batch. If the error increases => we know that this feature is important. If the error remains about the same => we know that the feature is not that important since the model was not very reliant on it. Finally, if the error decreases => we know likely know that the feature was negatively impacting the model's performance and hence was detrimental.

error/loss functions must be provided in the `forward_func`. See the test cases for examples.

### CODE REVIEW

**Comment by NarineK** [comment]
Thank you, @miguelmartin75 ! Would you, please, also add a brief description of the PR before the TODO list ?

File: captum/attr/_core/perm_feature_importance.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+
+class PermutationFeatureImportance(FeatureAblation):
+    def __init__(self, forward_func=None, perm_fn=permute_feature):
+        super().__init__(forward_func=forward_func)
+        self.perm_fn = perm_fn
+
+    def attribute(
+        self,
+        inputs=None,
```
Comment (NarineK) [inline captum/attr/_core/perm_feature_importance.py]: We assume that input must be provided in all attribution algorithms, meaning that we can have `inputs` without defaulting it to `None`

File: tests/attr/test_perm_feature_importance.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x, y):
+        return x[:, 0] * x[:, 1] + y[:, 0] * y[:, 1]
+
+
+class Test(BaseTest):
+    def test_perm_fn(self):
+        for bs in [2, 10, 100]:
```
[the remark this answers is not in the document]
Comment (miguelmartin75) [reply tests/attr/test_perm_feature_importance.py]: Should I simply assert, or throw an exception in this case? If I assert there will be no need to add a test case.

**Comment by NarineK** [comment]
Thank you for the update, @miguelmartin75 ! I'll take a look. It looks like the linter is failing. Also, it would be good to add a short description, in case someone in the open source community opens this PR(even after merging or closing) and wants to understand what it is about described in couple sentences.

File: tests/attr/test_perm_feature_importance.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        self.assertTrue(attribs[1].squeeze(0).size() == input_size)
+
+        for i in range(2):
+            self.assertTrue((attribs[0][:, i] != 0).all())
+            if i != 1:
+                self.assertTrue((attribs[1][:, i] != 0).all())
+
+        self.assertTrue((attribs[1][:, 1] == 0).all())
+
+    def test_mulitple_ablations(self):
```
Comment (NarineK) [inline tests/attr/test_perm_feature_importance.py]: nit: multiple permutations, perhaps ? Because we are testing permutation.

File: tests/attr/test_perm_feature_importance.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            unpermuted_features = mask.bitwise_not()
+            self.assertTrue((inp[:, mask] != perm_inp[:, mask]).any())
+            self.assertTrue(
+                (inp[:, unpermuted_features] == perm_inp[:, unpermuted_features]).all()
+            )
+
+        sizes_to_test = [(10,), (4, 5), (3, 4, 5), (6, 7, 8, 9)]
+        for batch_size in [2, 10]:
+            for inp_size in sizes_to_test:
+                inp = torch.randn((batch_size,) + inp_size)
```
Comment (NarineK) [inline tests/attr/test_perm_feature_importance.py]: The nested for -loops might lead to test timeouts. It is better to avoid nested for-loops in the tests and instead split it up into more tests.

File: tests/attr/test_perm_feature_importance.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                flat_mask = torch.zeros_like(inp[0]).flatten().bool()
+
+                # test random set of single features
+                num_features = inp.numel() // batch_size
+                num_features_to_test = min(num_features, random.randint(2, 30))
```
Comment (NarineK) [inline tests/attr/test_perm_feature_importance.py]: it is better to use a fixed number instead of `random.randint(2, 30)`

File: tests/attr/test_perm_feature_importance.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+                    self.assertTrue(torch.sum(flat_mask) == 1)
+
+                    # permute the feature
+                    mask = flat_mask.view_as(inp[0])
+                    perm_inp = permute_feature(inp, mask)
+                    check_features_are_permuted(inp, perm_inp, mask)
+
+                    flat_mask[feature_idx] = 0
+
+                # test random set of features
```
Comment (NarineK) [inline tests/attr/test_perm_feature_importance.py]: This could be a separate test.

File: tests/attr/test_perm_feature_importance.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+class Test(BaseTest):
+    def test_perm_fn(self):
+        def check_features_are_permuted(inp, perm_inp, mask):
+            unpermuted_features = mask.bitwise_not()
+            self.assertTrue((inp[:, mask] != perm_inp[:, mask]).any())
+            self.assertTrue(
+                (inp[:, unpermuted_features] == perm_inp[:, unpermuted_features]).all()
+            )
+
+        sizes_to_test = [(10,), (4, 5), (3, 4, 5), (6, 7, 8, 9)]
```
Comment (NarineK) [inline tests/attr/test_perm_feature_importance.py]: Can we also add a separate test case that calls `PermutationFeatureImportance(...).attribute(...) ` for the len(input.shape) > 2 ?

File: tests/attr/test_perm_feature_importance.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        for masks in masks_to_test:
+            target = random.randint(0, 2)
+            attribs = feature_importance.attribute(
+                (inp1, inp2), feature_mask=masks, target=target
+            )
+            self.assertTrue(isinstance(attribs, tuple))
+
+            features_in_mask = set()
+            for mask in masks:
+                fvs = mask
```
Comment (NarineK) [inline tests/attr/test_perm_feature_importance.py]: nit: It would be good to simplify this test case

File: captum/attr/_core/perm_feature_importance.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    assert n > 1, "cannot permute features with batch_size = 1"
+
+    perm = torch.randperm(n)
+    no_perm = torch.arange(n)
+    while (perm == no_perm).all():
+        perm = torch.randperm(n)
+
+    out = x.clone()
+    for i, j in enumerate(perm):
+        out[i, feature_mask] = x[j, feature_mask]
```
[the remark this answers is not in the document]
Comment (miguelmartin75) [reply captum/attr/_core/perm_feature_importance.py]: Using arithmetic now, but my code will require torch >=1.4 (maybe >=1.3) unless I convert the mask to the same dtype as the input

[the remark this answers is not in the document]
Comment (miguelmartin75) [reply captum/attr/_core/perm_feature_importance.py]: Using a .to in order for tests to pass (for older versions of pytorch)
